### PR TITLE
Refactor: decouple webhook processing service

### PR DESCRIPTION
**Pull request by htdat**
A part of this PR #3827

From this discussion here https://github.com/Automattic/woocommerce-payments/pull/3827#discussion_r812854891, we decided to decouple the current webhook controller to a separate service. 

The new service will be used to process data in the current webhook controller as well as the new webhook reliability service in PR #3827

#### Changes proposed in this Pull Request

- Decouple the processing part of `WC_REST_Payments_Webhook_Controller` to a separate service `WC_Payments_Webhook_Processing_Service`: 
     - Move the processing part in WC_REST_Payments_Webhook_Controller::handle_webhook() to the new service. 
     - Change the REST-specific exception `Rest_Request_Exception` to another webhook-specific exception `Invalid_Webhook_Data_Exception`. 
     - Change name `read_rest_property()` to `read_webhook_property()`
     - Move the code handling event type `wcpay.notification`. to a new private method `process_wcpay_notification`: convert the exception here. 

- Create a new test class for the service above: WC_Payments_Webhook_Processing_Service_Test
     - Move all tests from WC_REST_Payments_Webhook_Controller_Test to this test.
     - Update tests to not rely on the rest response. Instead, focus on the mock and Invalid_Webhook_Data_Exception to decide if all webhook processes are run properly.

- Add new tests for WC_REST_Payments_Webhook_Controller_Test to focus on the behavior of WC_REST_Payments_Webhook_Controller::handle_webhook().

#### Testing instructions

* Mostly check tests and confirm they're all passed. 
* If possible, create some webhook events and confirm there is no error in the PHP error log. 

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [x] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) : 'QA Testing Not Applicable'

### CODE REVIEW

File: includes/class-wc-payments-webhook-processing-service.php [hunk 1]
```diff
@@ -0,0 +1,511 @@
...
+	/**
+	 * Safely get a value from the REST request body array.
+	 *
+	 * @param array  $array Array to read from.
+	 * @param string $key   ID to fetch on.
+	 *
+	 * @return string|array|int
+	 * @throws Rest_Request_Exception Thrown if ID not set.
+	 */
+	public function read_rest_property( $array, $key ) {
```
Comment (RadoslavGeorgiev) [inline includes/class-wc-payments-webhook-processing-service.php:124]: In the context of a webhook processing service we should probably avoid REST terminology and classes (exceptions in this case).

Renaming the method would be the easy part (ex. `read_webhook_property`), and I think we could throw a new type of exception here, which can be converted to a `Rest_Request_Exception` in the webhook controller. WDYT?

**Review by RadoslavGeorgiev — COMMENTED** [review]
I think this is a good direction for decoupling webhooks from the controller. So far I've got a single comment.

File: includes/class-wc-payments-webhook-processing-service.php (hunk 1 above)
Comment (htdat) [reply includes/class-wc-payments-webhook-processing-service.php:124]: Agreed. I will iterate on this.

Comment (htdat) [reply includes/class-wc-payments-webhook-processing-service.php]: The latest version of the PR should have this already. It's almost ready for the review. I will double check again and request an official review next week.

**Comment by htdat** [comment]
Updated the PR code and description. Ready for review.

**Comment by htdat** [comment]
Hi @Automattic/helix, would love to have your review too as it touches some subscription code base.

**Review by mattallan — APPROVED** [review]
> would love to have your review too as it touches some subscription code base.

Thanks for the heads up on these changes @htdat ! I've performed a code review on this with a close eye on the WCPay Subscriptions event handling changes and everything looks good from that perspective.

I haven't been able to test WCPay Subscription specific webhooks are still working as we're waiting on our accounts to be flagged for a Stripe feature that lets us trigger `invoice.paid` event etc. That said, from a code review this looks good! Thanks

**Review by RadoslavGeorgiev — APPROVED** [review]
Excellent work, @htdat! 👏 

I don't have any further comments and everything seems to work like a charm! :shipit:

**Comment by haszari** [comment]
> I haven't been able to test WCPay Subscription specific webhooks are still working as we're waiting on our accounts to be flagged for a Stripe feature that lets us trigger invoice.paid event etc. That said, from a code review this looks good! Thanks

Just flagging this @htdat  – some members of Helix could test those flows, as it requires [Stripe billing clocks aka test clocks](https://stripe.com/docs/billing/testing/test-clocks) (not everyone has this enabled on their account). So if you would like the subscriptions-related flows tested before merge or ship we can make that happen – let me know or jump in slack. I think @aprea has test clocks working and could test.

Side note: I think test clocks is going out of beta soon (so should be available to all).

**Comment by htdat** [comment]
> Just flagging this @htdat – some members of Helix could test those flows, as it requires [Stripe billing clocks aka test clocks](https://stripe.com/docs/billing/testing/test-clocks) (not everyone has this enabled on their account). So if you would like the subscriptions-related flows tested before merge or ship we can make that happen – let me know or jump in slack. I think @aprea has test clocks working and could test.

Asking here p1646283049352659-slack-C02BW3Z8SHK

Temporarily add `status: blocked` for this PR to wait for the testing with subscriptions.

**Comment by mattallan** [comment]
Hey @htdat, I've been able to test this PR with the Subscription webhooks and it's all working flawlessly 💯

**Comment by htdat** [comment]
Merging this PR now so that it can go through all GlobalStep test cycles for WCPay 3.9.0.

**Pull request merged**